File: 0402-remove-k-digits/0402-remove-k-digits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function(num, k) {
    let stack = [], i = 0;

	while (i < num.length) {
        while(k>0 && stack.length && stack[stack.length-1] > num[i]) {
            stack.pop();
            k--;
        }

        stack.push(num[i]);
        i++;
    }
    while(k-- && stack.length ) {
        stack.pop();
    }

    while( stack[0] === '0' && stack.length ) {
        stack.shift();
    }
    
    return stack.length ? stack.join('') : '0';
};

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []

        for current in num:
            while stack and stack[-1] > current and k:
                stack.pop()
                k -= 1
            
            if not stack and current == "0":
                continue
                
            stack.append(current)
        
        while stack and k:
            stack.pop()
            k -= 1
        
        return "".join(stack) if stack else "0"
